example/gosumtype_1_declaration.go

```diff
@@ -15,22 +15,13 @@ import (
 //	    = CreditCard String Time
 //	    | Paypal String
 //
-// we declare `User` as an interface
-type User interface {
-	Match(s UserVariants)
-}
-
-// and the variants as a struct
+// if we declare the `UserVariants` as a struct, we'll get the `User` type.
 type UserVariants struct {
```

Comment:
**Biggest change** is that application code authors do NOT define the primary `interface` anymore; just continue to define the `*Variants struct`

The reason for this change is this error:

```
json: cannot unmarshal object into Go value of type User
```

while we can

```diff
	type User interface {
		Match(s UserVariants)
+		json.Marshaler
+		json.Unmarshaler
	}
```

and make any User value marshal into json, there is no way we can leverage on Go's "encoding/json" package to automatically decode a piece of json string back into a variant of User, aka `Anonymous` or `Member`. Go has to decode into a concrete type.

the solution is: change `User` into a struct so we can attach `MarshalJSON` and `UnmarshalJSON` methods. this makes all sum types support json

for this to happen, sumtype-go generator have to take over the generating of `type User struct`